src/ledger/database/lmdb.zig

```diff
@@ -0,0 +1,716 @@
+                        start_operation = .prev;
+                    },
+                }
+            } else if (direction == .reverse) {
+                _ = try cursorGetRelative(cursor, .last);
+            }
+
+            return .{
+                .allocator = self.allocator,
+                .txn = txn,
+                .cursor = cursor,
+                .direction = switch (direction) {
+                    .forward => .next,
+                    .reverse => .prev,
+                },
+                .next_operation = start_operation,
+            };
+        }
+
+        pub fn Iterator(cf: ColumnFamily, _: IteratorDirection) type {
+            return struct {
+                allocator: Allocator,
+                txn: *c.MDB_txn,
+                cursor: *c.MDB_cursor,
+                direction: CursorRelativeOperation,
+                next_operation: CursorRelativeOperation,
+
+                /// Calling this will free all slices returned by the iterator
+                pub fn deinit(self: *@This()) void {
+                    c.mdb_cursor_close(self.cursor);
+                    c.mdb_txn_abort(self.txn);
+                }
+
+                pub fn next(self: *@This()) anyerror!?cf.Entry() {
+                    const key, const val = try self.nextImpl() orelse return null;
+                    return .{
+                        try key_serializer.deserialize(cf.Key, self.allocator, key),
+                        try value_serializer.deserialize(cf.Value, self.allocator, val),
+                    };
+                }
+
+                pub fn nextKey(self: *@This()) anyerror!?cf.Key {
+                    const key, _ = try self.nextImpl() orelse return null;
+                    return try key_serializer.deserialize(cf.Key, self.allocator, key);
+                }
+
+                pub fn nextValue(self: *@This()) anyerror!?cf.Value {
+                    _, const val = try self.nextImpl() orelse return null;
+                    return try key_serializer.deserialize(cf.Value, self.allocator, val);
+                }
+
+                /// Returned data does not outlive the iterator.
+                pub fn nextBytes(self: *@This()) LmdbError!?[2]BytesRef {
+                    const key, const val = try self.nextImpl() orelse return null;
+                    return .{
+                        .{ .allocator = null, .data = key },
+                        .{ .allocator = null, .data = val },
+                    };
+                }
+
+                fn nextImpl(self: *@This()) LmdbError!?struct { []const u8, []const u8 } {
+                    defer self.next_operation = self.direction;
+                    return try cursorGetRelative(self.cursor, self.next_operation);
+                }
+            };
+        }
+    };
+}
+
+fn toVal(bytes: []const u8) c.MDB_val {
+    return .{
+        .mv_size = bytes.len,
+        .mv_data = @constCast(@ptrCast(bytes.ptr)),
```

Comment:
I dunno about this. While I see what you're doing with this constCast, it can lead to very illegal behavior. One of the *only* use cases of constCast is to mitigate incorrect APIs, which I actually don't think this is. 
I bet the implementation stores a mutable pointer because it frees it at some point. I have not looked into lmdb, so don't quote me on that.

Please take a `[]u8` here and duplicate and constant strings or string literals instead. I refuse to use constCast anywhere in this codebase.

Reply:
> Please take a []u8 here and duplicate and constant strings or string literals instead.

This seems doable. I probably won't need to dupe anything. It will require some changes to preexisting code, but shouldn't be too bad.

--------

> I bet the implementation stores a mutable pointer because it frees it at some point.

toVal is used for inputs to LMDB. These are pointers that are allocated by the caller (in zig) and passed into an LMDB function as an input. They are not freed by LMDB.

> One of the only use cases of constCast is to mitigate incorrect APIs, which I actually don't think this is.

My understanding is that when MDB_val is passed as an input into LMDB, the data behind the pointer is never mutated. It only reads the data.

Would you call that an incorrect API? If it says it takes mutable pointers but never mutates them?

> it can lead to very illegal behavior

What kind of illegal behavior could happen in this situation?

Reply:
Yeah lmdb's api is a bit odd w.r.t parameters there. As far as I understand it, yes, it will not modify that value at most call sites. The reason for this is some functions & options will use the key as an out param, e.g. `mdb_cursor_get` with `MDB_GET_CURRENT` will write to both the key and value params.

If you want to assert that it's const, I think it's best to be super explicit at the callsite of toVal. We wouldn't want to e.g. assert that something's const when it isn't (even if the memory is safely writeable).

Reply:
Weighing in:
> Would you call that an incorrect API? If it says it takes mutable pointers but never mutates them?

Yeah, this is explicitly an incorrect API; it doesn't adequately communicate its intent.

Reply:
Thanks for explaining everyone, I agree lmdb is in the wrong here. Nevertheless, we should not use `@constCast` if we can avoid it.

> What kind of illegal behavior could happen in this situation?

I originally assumed that either 
1. lmdb is freeing the data at some point, since free takes a mutable pointer. freeing a piece of data in a rodata section is invalid and would result in a panic.
2. it's writing to the data somehow. writing to data in rodata is also invalid and would cause a fault.

Reply:
Changing this to require a mutable pointer without adding any extra allocations is doable but awkward. It requires some extra complexity to distinguish between input types, which would need to be mutable, and output types, which should probably be immutable.

It's strange to require callers to always pass mutable pointers into the database, even though it will never be mutated. It might encourage callers to allocate dupes before passing it in, which would be pointless.